test/test_parser/test_parser.py

```diff
@@ -0,0 +1,665 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import sys
+import random
+import unittest
+
+from mathics.core.parser.errors import ScanError, IncompleteSyntaxError, InvalidSyntaxError
+from mathics.core.parser.ast import Node, Symbol, Number, String, Filename
+from mathics.core.parser.feed import SingleLineFeeder
+from mathics.core.parser.parser import Parser
+
+import six
+
+
+class ParserTests(unittest.TestCase):
+    def setUp(self):
+        self.parser = Parser()
+
+    def parse(self, s):
+        return self.parser.parse(SingleLineFeeder(s))
+
+    def check(self, expr1, expr2):
+        if isinstance(expr1, six.string_types):
+            expr1 = self.parse(expr1)
+        if isinstance(expr2, six.string_types):
+            expr2 = self.parse(expr2)
+
+        if expr1 is None:
+            self.assertIsNone(expr2)
+        else:
+            self.assertEqual(repr(expr1), repr(expr2))
+            self.assertEqual(expr1, expr2)
+
+    def scan_error(self, string):
+        self.assertRaises(ScanError, self.parse, string)
+
+    def incomplete_error(self, string):
+        self.assertRaises(IncompleteSyntaxError, self.parse, string)
+
+    def invalid_error(self, string):
+        self.assertRaises(InvalidSyntaxError, self.parse, string)
+
+
+class PrecedenceTests(ParserTests):
+    def test_minuslike(self):
+        self.check('a * + b', 'Times[a, Plus[b]]')
+        self.check('- a . b', 'Times[-1, Dot[a, b]]'),
+        self.check('- a / b', 'Times[-1, a, Power[b, -1]]'),
+        self.check('- a / - b', 'Times[-1, a, Power[Times[-1, b], -1]]')
+        self.check('- a / - b', 'Times[-1, a, Power[Times[-1, b], -1]]')
+        self.check('a + b!', 'Plus[a, Factorial[b]]')
+        self.check('!a!', 'Not[Factorial[a]]')
+
+
+class AssocTests(ParserTests):
+    def test_right(self):
+        self.check('a ^ b ^ c', 'a ^ (b ^ c)')
+
+    def test_flat(self):
+        self.check('a + b + c', 'Plus[a, b, c]')
+
+    def test_left(self):
+        self.check('a /; b /; c', '(a /; b) /; c')
+
+    def test_Subtract(self):
+        self.check('a - b - c', 'Plus[a, Times[-1, b], Times[-1, c]]')
```

Comment:
That's a good question. The behaviour is correct, see the [wolfram language spec](https://reference.wolfram.com/language/tutorial/OperatorInputForms.html), and it's also what MMA does.

As to why this is part of the language I'm not completely sure but I have some ideas:
* Converting `Subtract[a, b]` to `Plus[a, Times[-1, b]]` saves implementing `Subtract.
* `Plus` and `Times` are flat operators and can, in principle, be multi-threaded. 
* Doing the conversion at parse time prevents the AST from getting too tall (reducing call stack depths when walking the tree).

For interest, something similar happens with `Divide`; `a / b` is converted to `Times[a, Power[b, -1]]`.